ingest/validation/validate_metadata.py

```diff
@@ -187,72 +178,181 @@ def validate_cells_unique(metadata):
     return valid
 
 
-def collect_cell_for_ontology(metadatum, row_data, metadata):
+def collect_cell_for_ontology(metadatum, row_data, metadata, array=False):
     """Collect ontology info for a single metadatum into CellMetadata.ontology dictionary
     """
-    local_errors = set()
     logger.debug('Begin: collect_cell_for_ontology')
     if metadatum.endswith('__unit'):
         ontology_label = metadatum + '_label'
     else:
         ontology_label = metadatum + '__ontology_label'
-    try:
-        metadata.ontology[metadatum][
-            (row_data[metadatum], row_data[ontology_label])
-        ].append(row_data['CellID'])
-    except KeyError:
-        metadata.ontology[metadatum][(row_data[metadatum])].append(row_data['CellID'])
-    return local_errors
+    if array:
+        try:
+            ontology_dict = dict(zip(row_data[metadatum], row_data[ontology_label]))
+            for id, label in ontology_dict.items():
+                metadata.ontology[metadatum][(id, label)].append(row_data['CellID'])
+        except TypeError:
+            for id in row_data[metadatum]:
+                metadata.ontology[metadatum][(id)].append(row_data['CellID'])
+    else:
+        try:
+            metadata.ontology[metadatum][
+                (row_data[metadatum], row_data[ontology_label])
+            ].append(row_data['CellID'])
+        except KeyError:
+            metadata.ontology[metadatum][(row_data[metadatum])].append(
+                row_data['CellID']
+            )
+    return
 
 
-def collect_ontology_data(row_data, metadata):
+def collect_ontology_data(row_data, metadata, convention):
     """Collect unique ontology IDs for ontology validation
     """
     logger.debug('Begin: collect_ontology_data')
     for entry in row_data.keys():
-        if entry in metadata.type['convention']['ontology']:
-            # skip ontologies that are arrays for now
-            if entry in metadata.type['convention']['array']:
-                pass
-                # not collecting metadata for array-based metadata
+        if is_ontology_metadata(convention, entry):
+            if is_array_metadata(convention, entry):
+                collect_cell_for_ontology(entry, row_data, metadata, array=True)
             else:
                 collect_cell_for_ontology(entry, row_data, metadata)
     return
 
 
+def cast_boolean_type(value):
+    """Cast metadata value as boolean, if castable
+    """
+    if value.lower() == 'true':
+        return True
+    elif value.lower() == 'false':
+        return False
+    else:
+        return value
+
+
+def cast_integer_type(value):
```

Comment:
At first glance this struck me as over-engineering, but after reviewing more closely how you use it in e.g. line 269-271 this seems reasonable.

Reply:
Thanks - originally had both array handling and type casting in same if-elif and realized there were combinations of cases I was missing (and readability was terrible). This approach seemed to cover all cases. If there is a better way, I'm interested to know what other options I should consider next time.